### ISSUE TITLE
"Save" button in the LV2 Multi Sampler in Ardor 6 does not light up

### ISSUE THREAD

**Issue by inFlowia** [sadko4u/lsp-plugins#118]
If you make any changes to the saved preset, for example, change the pan or sample volume, this will not cause the "Save current preset" button to be activated (the Ardour button! not the internal button of the plugin). The only thing I found out of what can cause the button to activate is switching the workspace from the tools to the mixer. But even after the button has become active, it is better not to click on it, but about this [next bug report.](https://github.com/sadko4u/lsp-plugins/issues/119)

Ardour 6
Ubuntu Studio 19.10 64-bit
Multi Sampler x12 Stereo

**Comment by sadko4u** [comment]
I think it is a result of missing support of this feature: 

http://lv2plug.in/ns/ext/state#Changed

Since LSP Samplers do not show many ports as legacy LV2 parameters, such behaviour can be true.

**Comment by sadko4u** [comment]
**StateChanged** extension has been implemented and published around one year ago. So it's not surprise that I've missed it. Will add ASAP.

**Comment by sadko4u** [comment]
Today I've performed some code changes inside the LSP and here are my observations:
1. lv2:StateChanged even if reported to Ardour doesn't affect the behaviour of the button.
2. If we send patch:Set message to the plugin for changing some lv2:Path value, it doesn't affect the behaviour of the button.
3. The button triggers when we change legacy LV2 ports only.
4. Currently I see no way to notify the host about that fact that internal state of plugin has changed and it can be saved to a preset.

@x42 @drobilla @pauldavisthefirst
Any suggestions/recommendations?

**Comment by sadko4u** [comment]
All recent changes have been committed into **github-issue-118** branch.

**Comment by sadko4u** [comment]
Response from @x42, an Ardour developer:

> (12:56:43) rgareus: SadKo: StateChanged marks the session as dirty, but not the preset. same as in jalv currently
>(12:57:33) rgareus: SadKo: I think that's an oversight.
>(12:57:48) rgareus: SadKo: I'll fix it when I'm back home (Sunday)
>(12:58:47) rgareus: SadKo: so far only control-parameters (which are saved by the host) mark a preset as modified.

So we need to wait a bit for @x42's changes and test the behaviour on the nightly build of Ardour.

**Comment by x42** [comment]
That description above does indeed correctly describe the current situation. Except in 3. s/legacy port/control port/.

**Comment by sadko4u** [comment]
> That description above does indeed correctly describe the current situation. Except in 3. s/legacy port/control port/.

Thanks, @x42 
Will wait for changes to get nightly build of Ardour for tests.

**Comment by x42** [comment]
StateChanged marks the session as dirty, but not the preset. This is likely an oversight.
So far only control-parameters (which are saved by the host) mark a preset as modified.

**Comment by drobilla** [comment]
I'd be very happy to start referring to them as "legacy ports" :)

**Comment by x42** [comment]
Try Ardour 6.0.140  LV2's `patch_Set` as well as `state_StateChanged` Atom messages now both mark the session and preset as modified.

**Comment by sadko4u** [comment]
Tested on Ardour-6.0.140-dbg - still no effect.

In log I see that patch:Set has been received and processed by multi-sampler:
```
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/ui_ports.h: 663] write: writing patch event id=sf_0_5, path=/home/sadko/.hydrogen/data/drumkits/Forzee Stereo Drumkit/Crash18-3.wav (71)
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/ui_ports.h: 639] serialize: mapPath = (nil), path = /home/sadko/.hydrogen/data/drumkits/Forzee Stereo Drumkit/Crash18-3.wav
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h: 764] parse_atom_object: body->key (25) = http://lv2plug.in/ns/ext/patch#property
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h: 765] parse_atom_object: body->value.type (5) = http://lv2plug.in/ns/ext/atom#URID
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h: 770] parse_atom_object: body->value.body (747) = http://lsp-plug.in/plugins/lv2/multisampler_x12/ports#sf_0_5
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h: 764] parse_atom_object: body->key (26) = http://lv2plug.in/ns/ext/patch#value
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h: 765] parse_atom_object: body->value.type (2) = http://lv2plug.in/ns/ext/atom#Path
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h: 780] parse_atom_object: forwarding patch message to port sf_0_5
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/ports.h: 616] set_string: submitting path to '/home/sadko/.hydrogen/data/drumkits/Forzee Stereo Drumkit/Crash18-3.wav' (length = 72), flags=0x0
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h:1439] run: port changed: sf_0_5, value=0.000000
```

But the button doesn't activate (FYI: the URI http://lsp-plug.in/plugins/lv2/multisampler_x12/ports#sf_0_5 is not mapped to some 'legacy' LV2 control port).

Also, plugin reports the StateChange event to the host:
```
tail -f ardour.log | grep '#STATE'
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h: 435] change_state_atomic: #STATE state changed from 0 to 1
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h: 435] change_state_atomic: #STATE state changed from 1 to 2
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h:1109] transmit_atoms: #STATE MODE = 2
```
But also no change in the UI (button does not activate).
Short description: STATE = 0 means that internal state and state with host are in sync.
STATE = 1 means that state has changed but is still not reported to the host.
STATE = 2 means that state has changed and has been reported to the host (no more need to emit state:StateChanged event).

**Comment by sadko4u** [comment]
The issue with state:StateChanged has been solved since I've utilized invalid URL.
The issue with patch:Set is still not working.

**Comment by x42** [comment]
Thanks for the heads up!
`patch:Set` worked from generic UIs, but the message was not intercepted when created by custom GUIs. We've just fixed that in Ardour 6.0.141.

**Comment by sadko4u** [comment]
@x42 
Tested 6.0.143 version of Ardour. After sending patch:Set, the UI does not allow to overwrite the preset. Here's log of the plugin that received the patch:Set message:
```
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h: 770] parse_atom_object: body->key (25) = http://lv2plug.in/ns/ext/patch#property
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h: 771] parse_atom_object: body->value.type (5) = http://lv2plug.in/ns/ext/atom#URID
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h: 776] parse_atom_object: body->value.body (659) = http://lsp-plug.in/plugins/lv2/multisampler_x12/ports#mk_0_0
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h: 770] parse_atom_object: body->key (26) = http://lv2plug.in/ns/ext/patch#value
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h: 771] parse_atom_object: body->value.type (8) = http://lv2plug.in/ns/ext/atom#Float
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h: 786] parse_atom_object: forwarding patch message to port mk_0_0
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/ports.h: 307] pre_process: Port mk_0_0 has been changed, value=0.841395
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h:1424] run: port changed: mk_0_0, value=0.841395
[TRC][/home/sadko/eclipse/lsp-plugins/include/container/lv2/wrapper.h:1438] run: updating settings
```

**Comment by sadko4u** [comment]
I've just merged all related changes to the **devel** branch: ee16e0061931a06349e105b2f50e6c37fc4d57ce

List of changes:
* VST plugins now provide possibility to save and load presets and do not crash Host.
* Added support of lv2:StateChanged extension which properly works with Ardour 6.0.145 and higher.
* Refactored and simplified LV2 parameter transport between UI and plugin code.

So, @inFlowia , for your case it won't be fixed for Ardour 6.0 since @x42 also was required to implement some changes to Ardour. But you can download nightly build of Ardour and ensure that the problem has been fixed.

**Comment by sadko4u** [comment]
Available in 1.1.24!
